### PR TITLE
Fix vcat in case no data frames are passed

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/3074

### CODE REVIEW

File: test/dataframe.jl [hunk 1]
```diff
@@ -2049,7 +2052,7 @@ end
                                     cols=:orderequal)
 end
 
-@testset "vcat with source" begin
+@testset "vcat with source and reduce" begin
```
Comment (nalimilan) [inline test/dataframe.jl:2055]:
```suggestion
@testset "vcat with source and reduce(vcat, ...)" begin
```

File: test/dataframe.jl [hunk 2]
```diff
@@ -2066,12 +2071,30 @@ end
                               C=[fill(missing, 6); 7:9],
                               source=[-4, -4, -4, -3, -3, -3, -2, -2, -2])
         @test res.source isa CategoricalVector
+
+        res = reduce(vcat, [df1, df2, df3, df4], cols=:union, source=col => categorical(-4:-1))
+        @test res ≅ DataFrame(A=1:9, B=[1:6; fill(missing, 3)],
+                              C=[fill(missing, 6); 7:9],
+                              source=[-4, -4, -4, -3, -3, -3, -2, -2, -2])
+        @test res.source isa CategoricalVector
```
Comment (nalimilan) [inline test/dataframe.jl:2079]: Why not just do `isequal_coltyped(res, reduce(vcat, [df1, df2, df3, df4], cols=:union, source=col => categorical(-4:-1)))` in the block above?

**Comment by nalimilan** [comment]
I also noticed this weird behavior. AFAICT it's not addressed by this PR?
```julia
julia> vcat(DataFrame(), DataFrame())
0×0 DataFrame

julia> vcat(DataFrame(), DataFrame(), cols=[:a, :b])
0×0 DataFrame

julia> vcat(DataFrame(a=1:2), DataFrame(), cols=[:a, :b])
2×2 DataFrame
 Row │ a      b       
     │ Int64  Missing 
─────┼────────────────
   1 │     1  missing 
   2 │     2  missing 
```

File: test/dataframe.jl (hunk 2 above)
Comment (bkamins) [reply test/dataframe.jl:2079]: fixed - it was a leftover of old tests.

**Comment by bkamins** [comment]
> AFAICT it's not addressed by this PR?

It is addressed by this PR. I have added tests showing this and a comment why it is covered (essentially the reason is that in `_vcat` the `dfs` argument has `DataFrame()` data frames filtered-out).

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -1945,7 +1945,7 @@ end
 function _vcat(dfs::AbstractVector{AbstractDataFrame};
                cols::Union{Symbol, AbstractVector{Symbol},
                            AbstractVector{<:AbstractString}}=:setequal)
-
+    # note that DataFrame() objects are dropped from dfs before we call _vcat
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1948]:
```suggestion
    # note that empty DataFrame() objects are dropped from dfs before we call _vcat
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**